### DISCUSSION TITLE
New prop on the Screen component API: `element`

### DISCUSSION THREAD

**Discussion by tk-o** [react-navigation/react-navigation#10188] [Q&A] [answered]
I've been working on a cross-platform abstraction for React routing, and would like to understand if the Screen component API could get an extra `element` prop, which would take a React element (instead of a React component) so the composition capabilities would improve.

The justification for this idea was already proposed by the crew behind the React Router [here](https://reactrouter.com/docs/en/v6/upgrading/v5#advantages-of-route-element).

Using the `element` prop would override the `component `prop`, so these props should be mutually exclusive on a single Screen element.

How do you like this idea? Are there any issues around the `element` prop being introduced?
It would not be a breaking change, but rather an enhancement.

**Comment by satya164** [comment, marked as answer]
Unlike React Router, previous screens stay rendered in React Navigation. Passing a React Component allows React Navigation to reduce re-renders in inactive screens by applying `React.memo`.

Accepting an element makes it impossible to apply `React.memo` or other optimizations.

Though React Navigation also allows you to bypass this and have full flexibility by providing a render callback: https://reactnavigation.org/docs/screen/#children, so there's no need for an `element` API to do the same thing.

**Reply by tk-o** [reply]
Cheers, I've been comparing `react-router` and `react-navigation` today, and I figured that the underlying concepts are very similar in both of them, but there are several difference in the implementations.

`react-router` is a headless router, so it does not assume any fixed layout, while `react-navigation` has couple of fixed layouts available as navigators. Navigators would use `options` as props required by the fixed layouts, and screens would use `params`.

In `react-router`, a `<Route` element that has children would play a similar role as a `<Navigator` element in `react-navigation`. Both serve as some kind of a fixed layout for nested routes: `<Route` elements in `react-router` vs. `<Screen` elements in `react-navigation`).

**Reply by tk-o** [reply]
Another huge difference is that the `react-router` in the browser (or in any of its `<*Router` elements) would have a single _navigation history_, while `react-navigation` has such history kept on a `<Navigator` element level, and in fact, can have multiple _navigation history_ states present in the same time.

**Reply by tk-o** [reply]
Providing global navigation approach is possible via using refs to a particular navigator. You can export refs and the navigate functions from each Navigator. Example: https://reactnavigation.org/docs/navigating-without-navigation-prop
